docker_client_test.go

```diff
@@ -0,0 +1,41 @@
+package testcontainers
```

Comment:
I've done a quick test with current version of the main branch: I copied the `docker_client_test.go` file into that branch and run the tests with race detection, and the tests pass using the previous approach with sync.Once.

I'd appreciate more context on which situations they race condition occurs, to better understand the motivation for the change 🙏  Probably a code snippet or a link to the repro code will help me to with that.

Reply:
To make the added test fail it's enough to unset the `DOCKER_HOST` and then run it. It will fail (obviously) because the daemon is unreachable but it will also show a data race.

```
==================
WARNING: DATA RACE
Write at 0x0001010e9200 by goroutine 205:
  github.com/testcontainers/testcontainers-go.(*DockerClient).Info.func1()
      github.com/testcontainers/testcontainers-go/docker_client.go:41 +0xcc
  sync.(*Once).doSlow()
      sync/once.go:74 +0xb0
  sync.(*Once).Do()
      sync/once.go:65 +0x40
  github.com/testcontainers/testcontainers-go.(*DockerClient).Info()
      github.com/testcontainers/testcontainers-go/docker_client.go:40 +0x8c
  github.com/testcontainers/testcontainers-go.TestGetDockerInfo.func2.1()
      github.com/testcontainers/testcontainers-go/docker_client_test.go:34 +0xb0

Previous read at 0x0001010e9200 by goroutine 228:
  github.com/testcontainers/testcontainers-go.(*DockerClient).Info()
      github.com/testcontainers/testcontainers-go/docker_client.go:72 +0xc8
  github.com/testcontainers/testcontainers-go.TestGetDockerInfo.func2.1()
      github.com/testcontainers/testcontainers-go/docker_client_test.go:34 +0xb0

Goroutine 205 (running) created at:
  github.com/testcontainers/testcontainers-go.TestGetDockerInfo.func2()
      github.com/testcontainers/testcontainers-go/docker_client_test.go:32 +0xa4
  testing.tRunner()
      testing/testing.go:1595 +0x194
  testing.(*T).Run.func1()
      testing/testing.go:1648 +0x40

Goroutine 228 (finished) created at:
  github.com/testcontainers/testcontainers-go.TestGetDockerInfo.func2()
      github.com/testcontainers/testcontainers-go/docker_client_test.go:32 +0xa4
  testing.tRunner()
      testing/testing.go:1595 +0x194
  testing.(*T).Run.func1()
      testing/testing.go:1648 +0x40
```

Reply:
I can suggest that we try to capture that with a small test that does just that, WDYT?

```
	t.Run("unreachable docker daemon causes a gentle failure", func(t *testing.T) {
		t.Setenv("DOCKER_HOST", "")

		ctx := context.Background()
		c, err := NewDockerClientWithOpts(ctx)
		require.NoError(t, err)

		_, err = c.Info(ctx)
		require.Error(t, err)
	})
```

Reply:
Ah yes, that's because testcontainers can also extract the docker host from the context.

Well then let's not try to be too smart about it and let's not add the test above.

If you really want to see the failure then just unset both the env and context or use a dummy context e.g. 

```
docker context create dummy --docker "host=tcp://localhost:2376"
```

Reply:
Ok, I see. Just to understand it perfectly: this data race exists when the docker host is not available, right?

Reply:
Correct, albeit there might be more situations like e.g. network flakiness (docker host being available but intermittently) and so on.

Reply:
BTW, I'm able to reproduce the error with the above steps, thanks for the pointer.

> Correct, albeit there might be more situations like e.g. network flakiness (docker host being available but intermittently) and so on.

Cool, thanks for the explanation. So, with the previous implementation using sync.Once, there would be the case of any goroutine trying to access the do block at the moment it's reset ([old line 69](https://github.com/testcontainers/testcontainers-go/pull/1779/files#diff-9b3348e6532da95ddb19ea858dbceae344acf2988304efdbb1f2ea98ba4b03f3L69)), right?

Now I see it, and I understand the risks of resetting a once programatically 🤦 TBH when adding that part I felt "something" that I know realise.

Thanks in any case for adding light to it 👏

Reply:
> So, with the previous implementation using sync.Once, there would be the case of any goroutine trying to access the do block at the moment it's reset ([old line 69](https://github.com/testcontainers/testcontainers-go/pull/1779/files#diff-9b3348e6532da95ddb19ea858dbceae344acf2988304efdbb1f2ea98ba4b03f3L69)), right?

Yes, it was basically susceptible to a data race on that once variable.

Reply:
The CI is running, will approve this one right after the CI passes.

Thanks!